### PR TITLE
[RFC] Nblib nbnxn setup encapsulation

### PR DESCRIPTION
**Pull request by kanduri**
I managed to separate the user facing client code from the underlying GMX code using a translation layer.

The `ForceCalculator` object would contain in it two other objects. One of them would be the `GmxForceCalculator` which would work with all GMX data structures internally and the other would be a builder (`NbvSetupUtil`) for this that would take in the data structures created by the user and fill the required GMX structures underneath.

I also attempted to separate the data created in the beginning of the simulation and the data transformed in each step. 

I am yet to upload tests. I am still writing tests and modifying code to make it more testable. In the process, I am splitting the functions a bit more so that I can limit their scope for better unit testing.

### CODE REVIEW

**Comment by kanduri** [comment]
Some observations about the design:

- `GMXForceCalculator` knows nothing about the user level data structures and the simulation state. It is meant to evolve for the singular purpose of orchestrating internals of GROMACS.

- `NbvSetupUtil` (tentative naming, please suggest others) is the translation layer that holds a handle to the user level data structures and sets up the `GMXForceCalculator`

- Both of the above objects are structs with fully public interfaces and attributes. However, they are private members of `ForceCalculator`. User trying to script particle simulations cannot (and should not) be able to access it. However, we would be able to **test the functionality without using the client code**.

Looking brainstorm a bit on the tests we need to write and how we should invoke them.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+    ic.eeltype          = (options.coulombType == BenchMarkCoulomb::Pme ? eelPME : eelRF);
+    ic.coulomb_modifier = eintmodPOTSHIFT;
+    ic.rcoulomb         = options.pairlistCutoff;
+
+    // Reaction-field with epsilon_rf=inf
+    // TODO: Replace by calc_rffac() after refactoring that
+    ic.k_rf = 0.5 * std::pow(ic.rcoulomb, -3);
+    ic.c_rf = 1 / ic.rcoulomb + ic.k_rf * ic.rcoulomb * ic.rcoulomb;
+
```
Comment (sekelle) [inline src/gromacs/nblib/gmxsetup.cpp:199]: This is missing the most recent updates by Joe and me. Please merge the from nblib.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    put_atoms_in_box(PbcType::Xyz, box_, system_.coordinates());
+
+    nbnxn_atomdata_t*                nbat = gmxForceCalculator_p->nbv_->nbat.get();
+    gmxForceCalculator_p->verletForces_.resizeWithPadding(nbat->numAtoms());
+
+    return gmxForceCalculator_p;
+}
+
+GmxForceCalculator::GmxForceCalculator(SimulationState system, const NBKernelOptions &options) : enerd_(1, 0), verletForces_({})
```
Comment (sekelle) [inline src/gromacs/nblib/gmxsetup.cpp]: It's not so nice to copy the entire system just to access the box and the number of AtomTypes.
IMHO, neither the NbvSetupUtil, nor the GmxForceCalculator should have their own copy of the SimulationState. For now, I would say, we store a SimulationState in the ForceCalculator and pass on shared pointers to NbvSetupUtil and GmxForceCalculator.
We can then later change the ForceCalculator to also receive a shared pointer and make the user or mdrunner object own the SimulationState.

File: src/gromacs/nblib/gmxsetup.h [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    //! The massive class from which nbfp, shift_vec and ntypes would be used
+    t_forcerec forcerec_;
+
+    //! Tasks to perform in an MD Step
+    gmx::StepWorkload stepWork_;
+
+    explicit GmxForceCalculator(SimulationState system, const NBKernelOptions& options);
+
+    //! Contains array for computed forces
+    gmx::PaddedHostVector<gmx::RVec> verletForces_;
```
Comment (sekelle) [inline src/gromacs/nblib/gmxsetup.h:92]: I don't see the point of this member. The compute function returns by value, so the forces are first copied into the member, and then into the return value. This is worse compared to a function local object that's being returned by value in compute. The way to optimize this function would be to have compute take a reference and let the user supply the storage, but this we can address later.

File: src/gromacs/nblib/gmxsetup.cpp [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+    gmx::fillLegacyMatrix(system_.box().matrix(), box_);
+
+    gmxForceCalculator_p->forcerec_.nbfp  = nonbondedParameters_;
+    snew(gmxForceCalculator_p->forcerec_.shift_vec, SHIFTS);
+    calc_shifts(box_, gmxForceCalculator_p->forcerec_.shift_vec);
+
+    put_atoms_in_box(PbcType::Xyz, box_, system_.coordinates());
+
+    nbnxn_atomdata_t*                nbat = gmxForceCalculator_p->nbv_->nbat.get();
+    gmxForceCalculator_p->verletForces_.resizeWithPadding(nbat->numAtoms());
```
Comment (sekelle) [inline src/gromacs/nblib/gmxsetup.cpp]: This reintroduces the no-zero-init bug we just fixed

**Review by sekelle — CHANGES_REQUESTED** [review]

File: src/gromacs/nblib/gmxsetup.cpp (hunk 2 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.cpp]: Agreed. Turned it into a shared pointer for now.

File: src/gromacs/nblib/gmxsetup.cpp (hunk 1 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.cpp:199]: Done.

File: src/gromacs/nblib/gmxsetup.h (hunk 3 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.h:92]: Cool. Just returning a reference for now.

File: src/gromacs/nblib/gmxsetup.cpp (hunk 4 above)
Comment (kanduri) [reply src/gromacs/nblib/gmxsetup.cpp]: Fixed

**Pull request closed without merging**